File: src/main/java/net/easecation/ghosty/recording/player/updated/PlayerUpdatedOffhand.java
package net.easecation.ghosty.recording.player.updated;

import cn.nukkit.inventory.PlayerOffhandInventory;
import cn.nukkit.item.Item;
import cn.nukkit.utils.BinaryStream;
import net.easecation.ghosty.entity.PlaybackNPC;
import net.easecation.ghosty.recording.player.PlayerRecordNode;
import net.easecation.ghosty.util.PersistenceBinaryStreamHelper;

/**
 * Created by Mulan Lin('Snake1999') on 2016/11/19 17:02.
 * All rights reserved
 */
public class PlayerUpdatedOffhand implements PlayerUpdated {

    private Item item;

    public static PlayerUpdatedOffhand of(Item item) {
        return new PlayerUpdatedOffhand(item);
    }

    @Override
    public int getUpdateTypeId() {
        return PlayerUpdated.TYPE_OFFHAND;
    }

    @Override
    public boolean hasStates() {
        return true;
    }

    @Override
    public void processTo(PlaybackNPC ghost) {
        if (ghost != null && ghost.getOffhandInventory() != null) {
            PlayerOffhandInventory inv = ghost.getOffhandInventory();
            inv.setItem(item == null ? Item.get(Item.AIR) : item);
            inv.sendContents(ghost.getViewers().values());
        }
    }

    @Override
    public PlayerRecordNode applyTo(PlayerRecordNode node) {
        node.setOffhand(item);
        return node;
    }

    public PlayerUpdatedOffhand(BinaryStream stream, int formatVersion) {
        switch (formatVersion) {
            case 1, 2: {
                this.item = PersistenceBinaryStreamHelper.getItem(stream);
                break;
            }
            case 0: {
                read(stream);
                break;
            }
            default:
                throw new IllegalArgumentException("Unsupported format version: " + formatVersion);
        }
    }

    private PlayerUpdatedOffhand(Item item) {
        this.item = item;
    }

    @Override
    public boolean equals(Object obj) {
        if(!(obj instanceof PlayerUpdatedOffhand o)) return false;
        return (item.equals(o.item));
    }

    @Override
    public void write(BinaryStream stream) {
        PersistenceBinaryStreamHelper.putItem(stream, this.item);
    }

    @Override
    public void read(BinaryStream stream) {
        this.item = stream.getSlot();
    }

    @Override
    public String toString() {
        return "PlayerUpdatedOffhand{" +
            "item=" + item +
            '}';
    }
}

File: src/main/java/net/easecation/ghosty/recording/player/updated/PlayerUpdatedOffhand.kt
package net.easecation.ghosty.recording.player.updated

import cn.nukkit.item.Item
import cn.nukkit.utils.BinaryStream
import kotlinx.serialization.Serializable
import net.easecation.ghosty.entity.PlaybackNPC
import net.easecation.ghosty.recording.UpdateWithState
import net.easecation.ghosty.recording.player.PlayerRecordNode
import net.easecation.ghosty.serializer.NukkitItemSerializer
import net.easecation.ghosty.util.PersistenceBinaryStreamHelper

@Serializable
data class PlayerUpdatedOffhand(
    @Serializable(with = NukkitItemSerializer::class)
    val item: Item
) : PlayerUpdated, UpdateWithState {

    override fun getUpdateTypeId() = PlayerUpdated.TYPE_OFFHAND

    override fun processTo(ghost: PlaybackNPC) {
        val inv = ghost.offhandInventory ?: return
        inv.item = item
        inv.sendContents(ghost.viewers.values)
    }

    override fun applyTo(node: PlayerRecordNode) = node.also {
        it.offhand = this.item
    }

    companion object {
        @JvmStatic
        fun of(item: Item) = PlayerUpdatedOffhand(item)

        @JvmStatic
        val ADAPTER: PlayerUpdateAdapter<PlayerUpdatedOffhand> = Adapter
    }

    private object Adapter : PlayerUpdateAdapter<PlayerUpdatedOffhand> {
        override fun write(updated: PlayerUpdatedOffhand, stream: BinaryStream) {
            PersistenceBinaryStreamHelper.putItem(stream, updated.item)
        }

        override fun read(stream: BinaryStream): PlayerUpdatedOffhand = read(stream, 3)

        override fun read(stream: BinaryStream, version: Int): PlayerUpdatedOffhand {
            return when (version) {
                0 -> PlayerUpdatedOffhand(stream.slot)
                1, 2, 3 -> PlayerUpdatedOffhand(PersistenceBinaryStreamHelper.getItem(stream))
                else -> throw IllegalArgumentException("Unsupported format version: $version")
            }
        }
    }
}
